src/exberry_int/exberry_integration.py

```diff
@@ -0,0 +1,250 @@
+        self.logger.info(f"...OK")
+
+        create_session = {
+            'q': Endpoints.CreateSession,
+            'sid': 0,
+            'd': {
+                'apiKey': api_key,
+                'timestamp': time_str,
+                'signature': signature
+            }
+        }
+        await ws.send_json(create_session)
+        self.logger.info("Waiting for session request to be confirmed...")
+
+
+    async def __producer_coro(self, ws):
+        """ Send queued outbound messages through the Exberry websocket """
+        request_to_send = None
+        try:
+            while True:
+                await self.session_started.wait()
+                self.logger.debug("Awaiting next message to send...")
+                request_to_send = await self.__dequeue_outbound()
+                self.logger.info(f"Integration --> Exberry: {request_to_send}")
+                await ws.send_json(request_to_send)
+                request_to_send = None
+        except Exception as e:
+            self.logger.exception(f"Error in producer coroutine: {e}")
+        finally:
+            self.logger.info("Ending producer coroutine...")
+            if request_to_send:
+                self.logger.info('Message dequeued but not sent, prioritizing on reconnect...')
+                await self.enqueue_outbound(request_to_send, OutboundPriority.DEQUEUED_MESSAGE)
+
+
+    def __check_message_validity(self, msg: dict):
+        """
+        Ensure that only messages that require a ledger command are put on the
+        integration's queue
+        """
+        if not 'q' in msg: return False
+        endpoint = msg['q']
+
+        if endpoint == Endpoints.PlaceOrder or endpoint == Endpoints.CancelOrder:
+            return 'errorType' in msg or 'd' in msg and 'orderId' in msg['d']
+        elif endpoint == Endpoints.MassCancel:
+            return 'errorType' in msg or 'numberOfOrders' in msg['d']
+        elif endpoint == Endpoints.OrderBookDepth:
+            return 'errorType' in msg or msg['d']['messageType'] == 'Executed'
+        else:
+            return endpoint in Endpoints.ValidResponseEndpoints
+
+
+    async def __consumer_coro(self, ws):
+        """ Place incoming websocket messages onto the integration queue for processing in the main function """
+        try:
+            async for msg in ws:
+                if msg.type == WSMsgType.TEXT:
+                    msg_data = msg.json()
+                    if self.__check_message_validity(msg_data):
+                        self.logger.info(f"Exberry --> Integration: {msg_data}")
+                        await self.env.queue.put(msg_data)
+                    else:
+                        self.logger.debug(f'Message {msg_data} not necessary for contract creation, ignoring...')
+
+                elif msg.type == WSMsgType.ERROR:
+                    self.logger.error("Error message type! Breaking consumer loop...")
```

Comment:
Is there a message to be logged here that signifies the type of error?

Reply:
`WSMsgType.ERROR` does have a `data` field, I'll add that